File: dynamic programming/sprint-dp/linear/divisor.cpp
#include<bits/stdc++.h>
using namespace std;
long long intPow(long long base, int exp) {
    long long res = 1;
    while (exp > 0) {
        if (exp & 1) res *= base;
        base *= base;
        exp >>= 1;
    }
    return res;
}

vector<int> generatePrimes(int maxN){
    vector<bool> isPrime(maxN+1, true);
    isPrime[0] = isPrime[1] = false;
    for(int i=2; i*i<=maxN; i++){
        if(isPrime[i]){
            for(int j=i*i; j<=maxN; j+=i ){
                isPrime[j]= false;
            }
        }
    }
    vector<int> primes;
    for(int i=2; i<=maxN; i++){
        if(isPrime[i]) primes.push_back(i);
    }
    return primes;
}
void productInt(int n, vector<int>& subr, vector<vector<int>>& result){
    if(n==1){
        result.push_back(subr);
        return;
    }
    for(int i=2; i<=n; i++){
        if(subr.size()>=1 and i<subr.back()) continue;
        if(n%i==0){
            subr.push_back(i);
            productInt(n/i, subr, result);
            subr.pop_back();
        }
    }
}

long long minimumNum(int n){
    vector<int> subresult;
    vector<vector<int>> result;
    vector<int> primes = generatePrimes(100);
    productInt(n, subresult, result);
    long long ans= LLONG_MAX;
    for(auto nums: result){
        long long k=1, primeIndex=0;
        for(int i=nums.size()-1; i>=0; i--){
            k*= intPow(primes[primeIndex], nums[i]-1);
            primeIndex++; 
        }
        ans = min(ans, k);
    }
    return ans;

}
int main(){
    cout<<minimumNum(100);
}

File: dynamic programming/sprint-dp/linear/divisor.py
from math import sqrt

def generatePrimes(n):
    isPrime = [True]*(n+1)
    isPrime[0]=isPrime[1]=False
    for i in range(2, n+1):
        if isPrime[i]:
            for j in range(i*i, n+1, i):
                isPrime[j]=False
    primes = [i for i, prime in enumerate(isPrime) if prime]
    
    return primes

def productInt(n, subresult, result):
    if(n==1):
        result.append(subresult[:])
        return
    for i in range(2, n+1):
        if n%i==0:
            subresult.append(i)
            productInt(n//i, subresult, result)
            subresult.pop()

def isTriangulanr(k):
    n = int(sqrt(k*2))
    return ((n*(n+1))//2) == k

def minimum_num(n):
    primes = generatePrimes(100)
    product_int= [] 
    productInt(n, [], product_int)
    ans = float('inf')
    real_ans = 0
    for nums in product_int:
        k=1
        prime_index = 0
        for i in nums:
            k *= primes[prime_index] ** (i-1)
            prime_index+=1
        print("k==", k)
        if isTriangulanr(k):
            print(k)
            ans = min(ans, k)

    return ans

minimum_num(6)